[WIP] Fix and complete server-side JID validation and escaping.

Natureshadow opened the pull request.

Natureshadow committed:
Require Stringprep (from GitHub, for now).

Natureshadow committed:
Use Stringprep to validate JIDs.

mirabilos committed:
Fix php-stringprep inclusion.

Natureshadow committed:
Allow domain part to be an IP address.

Natureshadow committed:
Implement full domain part validation.

Natureshadow committed:
Escape all relevant characters in JIDs.

Natureshadow committed:
Try not to discard stuff in JID splitting.

Natureshadow committed:
Make echapJid only escape the local part.

Natureshadow committed:
Fix JID exploding and little mistake in escaping.

Natureshadow committed:
Use explodeJid in prepJid.

Natureshadow committed:
Allow all valid JIDs in search box.

edhelas commented:
Thanks a lo, I'll review it asap!
You've added a dependency, is it OK regarding the packages (like the Debian one?).

Natureshadow commented:
Hi,

> Thanks a lo, I'll review it asap!
> You've added a dependency, is it OK regarding the packages (like the Debian one?).

For Debian, I will of course take care of that.  For everything else, it has
to be - no JID validation is possible without that (except with copying the
code, of course).

edhelas commented on app/helpers/StringHelper.php:
```diff
@@ -131,7 +136,14 @@ function getHashtags($string): array
-    return str_replace(' ', '\40', $jid);
+    $jid_parts = explodeJid($jid);
+
+    $from = array('\\',   ' ',    '"',    '&',    '\'',   '/',    ':',    '<',    '>',    '@');
```
Looks great but I'll have to check if it doesn't break stuff in the UI

Natureshadow replied:
When displaying a JID, you SHOULD translate all the characters back, says the standard.

edhelas commented on app/helpers/StringHelper.php:
```diff
@@ -131,7 +136,14 @@ function getHashtags($string): array
+
+    $jid_parts['username'] = str_replace($from, $to, $jid_parts["username"]);
+
+    return ($jid_parts['username'] !== '' ? $jid_parts['username'].'@' : '').$jid_parts['server'].($jid_parts['resource'] !== '' ? '/'.$jid_parts['resource'] : '');
```
Could you please re-indent that part ? Like
```
$whatif
    ? 'yeah'
    : 'nope';
```

edhelas commented on app/helpers/StringHelper.php:
```diff
@@ -392,3 +412,86 @@ function urilize($path, bool $noTime = false): string
+/**
+ * Returns true if the value is an IPv6 or IPv4 address (in that order)
+ */
+function isIpAddress($value) {
```
Can you please put a carriage return to all the `function() {` in that file ? :3

Natureshadow replied:
Did that for everything I touched. Fixing that style for your own code should go outside this PR ;).

edhelas commented on app/helpers/StringHelper.php:
```diff
@@ -392,3 +412,86 @@ function urilize($path, bool $noTime = false): string
+        return false;
+    }
+
+    if ((strlen($ret) < 1) || (strlen($ret) > 1023)) {
```
1023, magic number ?

Natureshadow replied:
What do you mean?

edhelas commented on app/helpers/StringHelper.php:
```diff
@@ -392,3 +412,86 @@ function urilize($path, bool $noTime = false): string
+
+    // Validate local part against Nodeprep profile
+    if ($jid_parts['username'] !== '') {
+        $jid_parts['username'] = prepJidPart($jid_parts['username'], new Nodeprep());
```
I prefer to not put parenthesis if you don't have parameters (`new Nodeprep`). But that's ok like that :p

Natureshadow replied:
I learned on a Java type-writer :P

edhelas commented on app/helpers/StringHelper.php:
```diff
@@ -392,3 +412,86 @@ function urilize($path, bool $noTime = false): string
+    }
+
+    // Finally, validate domain part (IPv6, IPv4, FQDN, in that order)
+    if (!isIpAddress($jid_parts['server'])) {
```
Please be consistent, could you change all the variables to camelCase instead of snake_case. 
I prefer to use camelCase in Movim :)

edhelas commented on app/widgets/Chat/Chat.php:
```diff
@@ -1014,7 +1012,7 @@ private function prepareComposeList(array $list)
     private function validateJid($jid)
     {
-        return (Validator::stringType()->noWhitespace()->length(6, 60)->validate($jid));
+        return prepJid($jid);
```
You can simply remove that method and directly call prepJid ;) Same with the others.

Natureshadow committed:
Revert "Allow all valid JIDs in search box."

That was a bad idea as it breaks full-text search in posts, as every normal
word is a valid JID.

This reverts commit b35249b984e4dd6693119eba06cceba2ddfee41f.

Natureshadow committed:
Implement implodeJid function and use it.

Natureshadow committed:
Fix variable names.

Natureshadow committed:
Fix braces on function definitions.

Natureshadow committed:
Remove all validateJid methods and call prepJid directly.

Natureshadow commented:
I update a lot, please review again, @edhelas ☺!

Natureshadow committed:
Use explodeJid everywhere a JID is exploded.

No need to explode all over the place.

Natureshadow committed:
Account for some code passing null into explodeJid.

Natureshadow committed:
Use current variant of IDNA.

Natureshadow committed:
Fix usage of bare JIDs in messages.

Natureshadow committed:
Correctly rebuild bare JIDs in explodeJid.

Natureshadow committed:
Make HTML IDs really unique.

The previous code could lead to data from different users being handled for
the wrong UI element.  For example, from the view of this code, the JIDs
"foo-bar@example.com", "foo_bar@example.com", "foo.bar@example.com", which
are all distinct and valid JIds, were handled as one and the same.

Natureshadow committed:
Align IDs of HTML elements to use _ in sufficēs.

Previously, the use of - and _ for sufficēs was mixed, sometimes even for
the same names, and the elements were only matche due to the old, non-unique
cleanupId code turning _ into -.

edhelas commented:
As I'm not planning to maintain a specifical stringprep dependency from that I'm assuming that the only proper way to do so would be to directly call "idn" (http://www.gnu.org/software/libidn/manual/html_node/Invoking-idn.html) and call it from PHP. I'm already using it during the registration process (see https://api.movim.eu/accounts/register) and it has a complete Nodeprep implementation.

edhelas commented:
Let's see if I can revive this PR :)
